pulser-core/pulser/sequence/_seq_drawer.py

```diff
@@ -44,100 +46,107 @@
         # List of interpolation points
         interp_pts: defaultdict[str, list[list[float]]] = defaultdict(list)
         target: dict[Union[str, tuple[int, int]], Any] = {}
         for slot in sch:
             if slot.ti == -1:
                 target["initial"] = slot.targets
-                time += [0]
-                amp += [0.0]
-                detuning += [0.0]
-                phase += [0.0]
                 continue
-            if slot.type in ["delay", "target"]:
-                time += [
-                    slot.ti,
-                    slot.tf - 1 if slot.tf > slot.ti else slot.ti,
-                ]
-                amp += [0.0, 0.0]
-                detuning += [0.0, 0.0]
-                phase += [phase[-1]] * 2
-                if slot.type == "target":
-                    target[(slot.ti, slot.tf - 1)] = slot.targets
+            if slot.type == "target":
+                target[(slot.ti, slot.tf - 1)] = slot.targets
+                continue
+            if slot.type == "delay":
                 continue
             pulse = cast(Pulse, slot.type)
-            if isinstance(pulse.amplitude, ConstantWaveform) and isinstance(
-                pulse.detuning, ConstantWaveform
-            ):
-                time += [slot.ti, slot.tf - 1]
-                amp += [float(pulse.amplitude[0])] * 2
-                detuning += [float(pulse.detuning[0])] * 2
-                phase += [float(pulse.phase) / (2 * np.pi)] * 2
-            else:
-                time += list(range(slot.ti, slot.tf))
-                amp += pulse.amplitude.samples.tolist()
-                detuning += pulse.detuning.samples.tolist()
-                phase += [float(pulse.phase) / (2 * np.pi)] * pulse.duration
-                for wf_type in ["amplitude", "detuning"]:
-                    wf = getattr(pulse, wf_type)
-                    if isinstance(wf, InterpolatedWaveform):
-                        pts = wf.data_points
-                        pts[:, 0] += slot.ti
-                        interp_pts[wf_type] += pts.tolist()
-
-        if time[-1] < total_duration - 1:
-            time += [time[-1] + 1, total_duration - 1]
-            amp += [0, 0]
-            detuning += [0, 0]
-            phase += [phase[-1] if len(phase) else 0] * 2
+            for wf_type in ["amplitude", "detuning"]:
+                wf = getattr(pulse, wf_type)
+                if isinstance(wf, InterpolatedWaveform):
+                    pts = wf.data_points
+                    pts[:, 0] += slot.ti
+                    interp_pts[wf_type] += pts.tolist()
+
         # Store everything
-        time.pop(0)  # Removes the -1 in the beginning
-        data[ch] = ChannelDrawContent(time, amp, detuning, phase, target)
-        if hasattr(seq, "_measurement"):
-            data[ch].measurement = seq._measurement
+        samples = sch.get_samples()
+        data[ch] = ChannelDrawContent(
+            samples.extend_duration(total_duration), target
+        )
```

Comment:
So it trims down to this thanks to the sampler, isn't it?

Reply:
Yes, exactly. In the end, it didn't make sense to have `ChannelSamples` make the `ChannelDrawContent`, but rather have give the `ChannelSamples` to `ChannelDrawContent`